### PR TITLE
enforce compatible pairings of domains/metrics

### LINKED ISSUE #185: Compatible pairings of domains and metrics
There is interest in using the type system to enforce compatible pairings of metrics and domains.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #185.

This PR also removes the domain-metric compatibility section from proofs.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
Current dependencies on/for this PR:
* main
  * **PR #648** <a href="https://app.graphite.dev/github/pr/opendp/opendp/648" target="_blank">Graphite</a> 
    * **PR #618** <a href="https://app.graphite.dev/github/pr/opendp/opendp/618" target="_blank">Graphite</a> 
      * **PR #674** <a href="https://app.graphite.dev/github/pr/opendp/opendp/674" target="_blank">Graphite</a> 
        * **PR #675** <a href="https://app.graphite.dev/github/pr/opendp/opendp/675" target="_blank">Graphite</a> 
          * **PR #676** <a href="https://app.graphite.dev/github/pr/opendp/opendp/676" target="_blank">Graphite</a> 
            * **PR #678** <a href="https://app.graphite.dev/github/pr/opendp/opendp/678" target="_blank">Graphite</a> 
              * **PR #682** <a href="https://app.graphite.dev/github/pr/opendp/opendp/682" target="_blank">Graphite</a> 
      * **PR #645** <a href="https://app.graphite.dev/github/pr/opendp/opendp/645" target="_blank">Graphite</a> 
        * **PR #604** <a href="https://app.graphite.dev/github/pr/opendp/opendp/604" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/604?utm_source=stack-comment).

File: rust/src/transformations/cast_metric/ffi.rs [hunk 1]
```diff
@@ -32,17 +32,20 @@ pub extern "C" fn opendp_transformations__make_ordered_random(
     where
         MI::Distance: 'static,
         <MI::OrderedMetric as Metric>::Distance: 'static,
+        (AnyDomain, MI): MetricSpace,
+        (AnyDomain, MI::OrderedMetric): MetricSpace,
     {
         let trans = try_!(super::make_ordered_random::<AnyDomain, MI>(domain));
 
-        FfiResult::Ok(util::into_raw(Transformation::new(
+        Transformation::new(
```
Comment (andrewvyrros) [inline rust/src/transformations/cast_metric/ffi.rs:40]: In hindsight, it's funny to see how much cleaner stuff like this becomes with the new flavor of `Transformation::new`.

File: rust/src/transformations/lipschitz_mul/ffi.rs [hunk 2]
```diff
@@ -37,15 +37,16 @@ pub extern "C" fn opendp_transformations__make_lipschitz_float_mul(
...
             make_lipschitz_float_mul::<D, M>(constant, bounds).into_any()
         }
 
         let constant = *try_as_ref!(constant as *const T);
         let bounds = try_!(try_as_ref!(bounds).downcast_ref::<(T, T)>());
         dispatch!(monomorphize2, [
-            (D, [AtomDomain<T>, VectorDomain<AtomDomain<T>>]),
-            (M, [AbsoluteDistance<T>, L1Distance<T>, L2Distance<T>])
+            (D, [AtomDomain<T>]),
+            (M, [AbsoluteDistance<T>])
```
Comment (andrewvyrros) [inline rust/src/transformations/lipschitz_mul/ffi.rs:49]: This is a nice little catch for `MetricSpace`!

**Review by andrewvyrros — APPROVED** [review]
This is another one that just feels right! I was stunned to see that the original discussion was so long ago. Thanks for pushing it forward!

The change to have `Transformation::new` and `Measurement::new` return `Fallible` is such a good call. It's interesting to me that we lived with the hassle of the old flavor for so long, I guess it just became background noise.

File: rust/src/combinators/chain/mod.rs [hunk 3]
```diff
@@ -292,17 +303,19 @@ mod tests {
...
+        let chain = make_chain_pm(&function1, &measurement0)?;
 
         let arg = 99_u8;
-        let ret = chain.invoke(&arg).unwrap_test();
+        let ret = chain.invoke(&arg)?;
         assert_eq!(ret, 101.0);
 
         let d_in = 99_i32;
-        let d_out = chain.map(&d_in).unwrap_test();
+        let d_out = chain.map(&d_in)?;
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/combinators/chain/mod.rs:315]: We can get rid of it everywhere. I didn't know about std::process::Termination at the time.

File: rust/src/core/mod.rs [hunk 4]
```diff
@@ -248,21 +254,25 @@ impl<DI: Domain, TO, MI: Metric, MO: Measure> Clone for Measurement<DI, TO, MI,
...
+{
     pub fn new(
         input_domain: DI,
         function: Function<DI::Carrier, TO>,
         input_metric: MI,
         output_measure: MO,
         privacy_map: PrivacyMap<MI, MO>,
-    ) -> Self {
-        Self {
+    ) -> Fallible<Self> {
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/core/mod.rs:267]: One less thing to explain to newcomers (+1), but one more error to handle (-1). Luckily most places we use it already want to return an error.

File: rust/src/transformations/cast_metric/ffi.rs (hunk 1 above)
Comment (Shoeboxam) [reply rust/src/transformations/cast_metric/ffi.rs:40]: It probably should have been `Ok(Transformation::new(...)).into()` before!

File: rust/src/transformations/count/make_count.tex [hunk 5]
```diff
@@ -80,16 +80,6 @@ \section{Proofs}
...
-\begin{proof} \textbf{(Part 2 -- domain-metric compatibility).}
-    Our \texttt{input\_metric} of \rustdoc{metrics/struct}{SymmetricDistance} is compatible with any domain of the form \texttt{VectorDomain(inner\_domain)}, 
-    and our \texttt{input\_domain} of \\\texttt{VectorDomain(AtomDomain(TIA))} is of this form. 
-    Therefore our \texttt{input\_domain} and \texttt{input\_metric} are compatible.
-
-    Our \texttt{output\_metric} of \rustdoc{metrics/struct}{AbsoluteDistance} is compatible with any domain of the form \texttt{AtomDomain(T)} where \texttt{T} has the trait \texttt{InfSub}, 
-    and our \texttt{output\_domain} of \texttt{AtomDomain(TO)} is of this form and \texttt{TO} has the necessary trait.
-    Therefore our \texttt{input\_domain} and \texttt{input\_metric} are compatible.
-\end{proof}
-
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/count/make_count.tex:92]: Get outta hea'!

**Pull request merged**